### PR TITLE
Make archival config more dynamic

### PR DESCRIPTION
**Pull request by yycptt**
https://github.com/uber/cadence/issues/2397

### CODE REVIEW

File: common/archiver/archivalMetadata.go [hunk 1]
```diff
@@ -148,34 +123,53 @@ func NewArchivalConfig(
 	}
 
 	return &archivalConfig{
-		clusterStatus:       clusterStatus,
-		enableRead:          enableRead,
-		domainDefaultStatus: domainDefaultStatus,
-		domainDefaultURI:    domainDefaultURI,
+		staticClusterStatus:  staticClusterStatus,
+		dynamicClusterStatus: dc.GetStringProperty(dynamicconfig.HistoryArchivalStatus, staticClusterStatusStr),
```
Comment (andrewjdawson2016) [inline common/archiver/archivalMetadata.go]: NewArchivalConfig seems like it is meant to work for both visibility and history but as its written this won't work for both?

File: common/archiver/archivalMetadata.go [hunk 2]
```diff
@@ -148,34 +123,53 @@ func NewArchivalConfig(
...
+	// even dynamic config says archival is enabled, we should ignore that.
+	// Only when archival is enabled in static config, should we check if there's any difference between static config and dynamic config.
+	if a.staticClusterStatus != ArchivalEnabled {
+		return a.staticClusterStatus
+	}
+
+	dynamicStatusStr := a.dynamicClusterStatus()
+	dynamicStatus, err := getClusterArchivalStatus(dynamicStatusStr)
+	if err != nil {
+		dynamicStatus = a.staticClusterStatus
```
Comment (andrewjdawson2016) [inline common/archiver/archivalMetadata.go]: I think that if we ever fail to read dynamic config we should always default to disabled state even if archival config is enabled. Thoughts?

File: common/archiver/archivalMetadata.go (hunk 1 above)
Comment (yycptt) [reply common/archiver/archivalMetadata.go]: Makes sense. I've updated the function to take dynamic function as parameter.

File: common/archiver/archivalMetadata.go (hunk 2 above)
Comment (yycptt) [reply common/archiver/archivalMetadata.go]: I can argue both ways and I don't have a strong opinion on this. so changed to disabled when dynamic config is invalid.

**Pull request merged**